### PR TITLE
Added Webhooks, Improved Cache

### PR DESCRIPTION
**Pull request by devgenix**

### CODE REVIEW

**Comment by devgenix** [comment]
@aybruhm skipping checks because I can't get it to use a valid host.

File: academia/signals.py [hunk 1]
```diff
@@ -1,13 +1,67 @@
...
+    cache_key = f"departments_{school_code}_{faculty_name}"
+    if cache.get(cache_key):
+        cache.delete(cache_key)
+        
+    # Dispatch webhook
+    webhook("department", instance.id)
+
+post_save.connect(invalidate_country_cache, sender=Country)
+post_save.connect(webhook_clear_school_cache, sender=School)
+post_save.connect(webhook_clear_faculty_cache, sender=Faculty)
```
Comment (aybruhm) [inline academia/signals.py:66]: I thought you've set up the server to support Redis and celery? If yes, is there a reason why you used Django signals? 

Django signals are not inherently asynchronous. As a matter of fact, it is a "blocking code". When a signal is triggered, the receivers (codes that are called in response to the signal) are executed synchronously, meaning that each request must complete before the next one can begin. 

Another caveat to using signals is the potential impact it'll have on scalability. When using synchronous signal handling, the application will have to wait for all the receivers to complete whatever they're doing before continuing with other requests (just like the default per-thread connection to the database with Django - Code blocking). 

Do not misunderstand me, man. Django signals are a powerful set of utility box that have helped, and will continue to help us in so many ways. However, if we're looking to start reducing the latency of our APIs, or reducing work load from our server, then we will need to start using asynchronous tasks with Celery and Redis (or RabbitMQ).

**Comment by aybruhm** [comment]
> @aybruhm skipping checks because I can't get it to use a valid host. 

Mock testing would handle that. But it's fine, though. I'll do it when I'm done with the feature you requested.

File: academia/signals.py (hunk 1 above)
Comment (aybruhm) [reply academia/signals.py:66]: Ah, damnit. I didn't realise until now that this repository is univast's. 

_Life of an engineer that decided to check his mail without putting on his glasses, smh._

It's all good, man. Cancel out the use of asynchronous tasks.

Comment (devgenix) [reply academia/signals.py:66]: > Ah, damnit. I didn't realise until now that this repository is univast's.
> 
> _Life of an engineer that decided to check his mail without putting on his glasses, smh._
> 
> It's all good, man. Cancel out the use of asynchronous tasks.

It's all good man. I understand. 
I wanted to get the webhook to be dispatched and working then implement celery and Redis to make it asynchronous.

Comment (devgenix) [reply academia/signals.py:66]: > I thought you've set up the server to support Redis and celery? If yes, is there a reason why you used Django signals?
> 
> Django signals are not inherently asynchronous. As a matter of fact, it is a "blocking code". When a signal is triggered, the receivers (codes that are called in response to the signal) are executed synchronously, meaning that each request must complete before the next one can begin.
> 
> Another caveat to using signals is the potential impact it'll have on scalability. When using synchronous signal handling, the application will have to wait for all the receivers to complete whatever they're doing before continuing with other requests (just like the default per-thread connection to the database with Django - Code blocking).
> 
> Do not misunderstand me, man. Django signals are a powerful set of utility box that have helped, and will continue to help us in so many ways. However, if we're looking to start reducing the latency of our APIs, or reducing work load from our server, then we will need to start using asynchronous tasks with Celery and Redis (or RabbitMQ).

I have now done that [here](https://github.com/faradayafrica/Univast/pull/23)

Comment (aybruhm) [reply academia/signals.py:66]: All good. Good job.

**Pull request merged**